### PR TITLE
Resteasy Reactive: Support produces Multipart FormData

### LINKED ISSUE #21009: Support MultiPart form result in Resteasy Reactive
### Describe the bug

It seems resteasy reactive cannot return a multipart body. In the reproducer below, I have a method like this

```java
	@GET
	@Path("/multipart")
	@Produces(MediaType.MULTIPART_FORM_DATA)
	public MultipartDto getMultipart() throws IOException {
		File file = File.createTempFile("abc", "txt");

		var writer = new FileWriter(file);
		writer.write("Test ABC");
		writer.close();

		return new MultipartDto("test text", file);
	}
```

But when I call it, I only get the following body:

```org.acme.MultipartDto@345fa2e``` 

### Expected behavior

Return a correct multipart form data response

### Actual behavior

Only toString of the returned object is returned.

### How to Reproduce?

1. Download reproducer from
[2021-10-26_restclient-reactive-multipart-response.zip](https://github.com/quarkusio/quarkus/files/7418137/2021-10-26_restclient-reactive-multipart-response.zip)
 2. Start quarkus with `./gradlew quarkusDev`
 3. Call GET localhost:8080/hello/multipart

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Sgitario**
Add RESTEasy Reactive support to produce Multipart Form data.
- Updated RESTEasy Reactive documentation
- Supported both `@RestForm` and `@FormParam` annotations
- Added test coverage for text/plan, json and xml serialization
- Implementation is based on Resteasy Classic documentation

Resolves https://github.com/quarkusio/quarkus/issues/21009

### CODE REVIEW

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    private static final String BOUNDARY_PARAM = "boundary";
+
+    // TODO: This needs to be configured using RuntimeConfiguration.body().defaultCharset()
+    private final Charset defaultCharset = StandardCharsets.UTF_8;
+
+    @Override
+    public void writeTo(Object o, Class<?> aClass, Type type, Annotation[] annotations, MediaType mediaType,
+            MultivaluedMap<String, Object> httpHeaders, OutputStream outputStream)
+            throws IOException, WebApplicationException {
+        if (o instanceof String) {
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: Does this writer still get used when the return type is string?

[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: No, I deleted this as the writer is not always configured, but only when the multipart form data is used.

[the remark this answers is not in the document]
Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: But my question still remains: When the writer is configured, is it being used to write out strings? If so, that is wrong.

File: docs/src/main/asciidoc/resteasy-reactive.adoc [hunk 2]
```diff
@@ -474,6 +472,49 @@ Moreoever if `quarkus.http.body.delete-uploaded-files-on-end` is set to true, Qu
...
+    @GET
+    @Produces(MediaType.MULTIPART_FORM_DATA)
+    @Path("file")
+    public DownloadFormData getFile() {
+        // return something
+    }
+}
+----
+
+WARNING: Returning Multipart data has been limited to be only working for blocking endpoints in order to avoid the event loop thread to deal with files.
```
Comment (geoand) [inline docs/src/main/asciidoc/resteasy-reactive.adoc:517]:
```suggestion
WARNING: For the time being, returning Multipart data is limited to be blocking endpoints. 
```

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: No, this is the final version of the writer: https://github.com/quarkusio/quarkus/pull/21038/files#diff-e8ce68a54c45a2a5550a9849588de76e36a5e0ba39530a2adab777d97978a20dR32

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/FormDataOutputMapperGenerator.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+     * The generated mapper would look like:
+     * public class FormData_generated_mapper implements MultipartOutputInjectionTarget {
+     * public FormDataOutput mapFrom(Object var1) {
+     * FormDataOutput var2 = new FormDataOutput();
+     * FormData var4 = (FormData)var1;
+     * File var3 = var4.data;
+     * MultipartSupport.addPartItemToFormDataOutput(var2, "file", "application/octet-stream", var3);
+     * File var5 = var4.text;
+     * MultipartSupport.addPartItemToFormDataOutput(var2, "text", "text/plain", var5);
+     * return var2;
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/FormDataOutputMapperGenerator.java:128]: This could use better formatting as it's not readable

File: independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java [hunk 4]
```diff
@@ -535,6 +542,12 @@ private ResourceMethod createResourceMethod(ClassInfo currentClassInfo, ClassInf
                 }
             }
 
+            if (returnsMultipart && !blocking) {
+                throw new DeploymentException(
+                        "Endpoints with '@Produces(MediaType.MULTIPART_FORM_DATA)' can only be used on non blocking methods. Offending method is '"
```
Comment (geoand) [inline independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java:547]:
```suggestion
                        "Endpoints that produce a Multipart result can only be used on non blocking methods. Offending method is '"
```

**Comment by geoand** [comment]
Let's only include this after https://github.com/quarkusio/quarkus/pull/21138 (which is very important) has gone in

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: And you have verified that returning a `String` continues to work properly even when this multipart writer has been configured do to it's use by some other method?

[the remark this answers is not in the document]
Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: Yes, I added this test case: https://github.com/quarkusio/quarkus/pull/21038/files#diff-136f284323fc8e3022c3f73dce4ed1d4e1ce5d4adf1c6e9affe221f6b0c6befdR48

[the remark this answers is not in the document]
Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: Great, thanks!

File: docs/src/main/asciidoc/resteasy-reactive.adoc (hunk 2 above)
Comment (Sgitario) [reply docs/src/main/asciidoc/resteasy-reactive.adoc:517]: Updated

**Comment by geoand** [comment]
There is a conflict now that https://github.com/quarkusio/quarkus/pull/21138 went in

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/FormDataOutputMapperGenerator.java (hunk 3 above)
Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/FormDataOutputMapperGenerator.java:128]: Ups, this is wrong. Thanks for spotting. Fixed.

File: independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java (hunk 4 above)
Comment (Sgitario) [reply independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java:547]: Updated

**Review by geoand — APPROVED** [review]
Thanks!

**Comment by geoand** [comment]
Something keeps breaking in CI, so I think it's an actual problem that needs to be fixed.

**Comment by Sgitario** [comment]
> Something keeps breaking in CI, so I think it's an actual problem that needs to be fixed.

The new writer was not registered for reflection and hence was failing when running on Native. Fixed.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java [hunk 5]
```diff
@@ -612,6 +613,9 @@ public void serverSerializers(ResteasyReactiveRecorder recorder,
         }
 
         serverSerializersProducer.produce(new ServerSerialisersBuildItem(serialisers));
+
+        // conditional serialisers
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, MultipartMessageBodyWriter.class.getName()));
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java:618]: Why are we always registering this for reflection even if it's not going to be used?

Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java:618]: I tried to register conditionally in here: QuarkusServerEndpointIndexer.handleMultipartForReturnType, but didn't work. Let me try again.

Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java]: It should be fixed now.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+public class MultipartMessageBodyWriter extends ServerMessageBodyWriter.AllWriteableMessageBodyWriter {
+
+    private static final String RESTEASY_DEFAULT_CHARSET_PROPERTY = "quarkus.resteasy-reactive.multipart.input-part.default-charset";
+    private static final Charset RESTEASY_DEFAULT_CHARSET_PROPERTY_DEFAULT = StandardCharsets.UTF_8;
+    private static final String DOUBLE_DASH = "--";
+    private static final String LINE_SEPARATOR = "\r\n";
+    private static final String BOUNDARY_PARAM = "boundary";
+
+    private Charset defaultCharset;
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java:40]: The way you are setting this, this needs to be `volatile`. 

But a better option would be to set it in the constructor, no?

**Review by geoand — CHANGES_REQUESTED** [review]
See my latest comment

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java (hunk 6 above)
Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: Changed

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+public class MultipartMessageBodyWriter extends ServerMessageBodyWriter.AllWriteableMessageBodyWriter {
+
+    private static final String RESTEASY_DEFAULT_CHARSET_PROPERTY = "quarkus.resteasy-reactive.multipart.input-part.default-charset";
+    private static final Charset RESTEASY_DEFAULT_CHARSET_PROPERTY_DEFAULT = StandardCharsets.UTF_8;
+    private static final String DOUBLE_DASH = "--";
+    private static final String LINE_SEPARATOR = "\r\n";
+    private static final String BOUNDARY_PARAM = "boundary";
+
+    private volatile Charset defaultCharset;
```
Comment (geoand) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java:40]: I guess I was not clear enough.

`volatile` is only needed if the field is **not** set in the constructor, but in a method that can be called by other threads.

If you set the field in the constructor, `volatile` is not needed and the field can be `final`.

Comment (Sgitario) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/multipart/MultipartMessageBodyWriter.java]: Ok, it makes more sense now. Changed.

**Review by geoand — APPROVED** [review]
